### PR TITLE
Mutex and RwLock arc method

### PR DESCRIPTION
**Pull request by MarinPostma**
This PR adds convenient methods `Mutex::arc` and `RwLock::arc` according to https://github.com/rust-lang/rust/issues/74866

Updates all the test to use the new methods instead of the old `Arc::new(..)`

### CODE REVIEW

File: library/std/src/sync/rwlock.rs [hunk 1]
```diff
@@ -136,6 +137,21 @@ impl<T> RwLock<T> {
             data: UnsafeCell::new(t),
         }
     }
+
+    /// Creates a new instance of an `RwLock<T>` which is unlocked, and wraps it in an [`Arc`].
```
Comment (pickfire) [inline library/std/src/sync/rwlock.rs:141]: Looks complicated. Why not just say this is a shorthand for `Arc<RwLock<T>>`?

File: library/std/src/sync/rwlock.rs [hunk 2]
```diff
@@ -136,6 +137,21 @@ impl<T> RwLock<T> {
...
+    ///
+    /// [`Arc`]: ../../std/sync/struct.Arc.html
```
Comment (pickfire) [inline library/std/src/sync/rwlock.rs:143]:
```suggestion
    ///
    /// [`Arc`]: ../../std/sync/struct.Arc.html
```

File: library/std/src/sync/rwlock.rs [hunk 3]
```diff
@@ -136,6 +137,21 @@ impl<T> RwLock<T> {
...
+    /// # Examples
```
Comment (pickfire) [inline library/std/src/sync/rwlock.rs:144]:
```suggestion
    ///
    /// # Examples
```

File: library/std/src/sync/mutex.rs [hunk 4]
```diff
@@ -229,6 +230,21 @@ impl<T> Mutex<T> {
         }
         m
     }
+
+    /// Creates a new mutex in an unlocked state ready for use, and wraps it in an [`Arc`].
+    ///
+    /// [`Arc`]: ../../std/sync/struct.Arc.html
+    /// # Examples
```
Comment (pickfire) [inline library/std/src/sync/mutex.rs:237]:
```suggestion
    ///
    /// # Examples
```

File: library/std/src/sync/mutex.rs [hunk 5]
```diff
@@ -229,6 +230,21 @@ impl<T> Mutex<T> {
         }
         m
     }
+
+    /// Creates a new mutex in an unlocked state ready for use, and wraps it in an [`Arc`].
```
Comment (pickfire) [inline library/std/src/sync/mutex.rs:234]: Sounds complicated. Why not just say this is a shorthand for `Arc::new(Mutex::new(t))` or this creates `Arc<Mutex<T>>`?

**Comment by pickfire** [comment]
Is the changes in submodule necessary?

File: library/std/src/sync/rwlock.rs (hunk 2 above)
Comment (jyn514) [reply library/std/src/sync/rwlock.rs:143]: Yes, since `Arc` is in scope this will be resolved by intra-doc links.

**Comment by MarinPostma** [comment]
> Is the changes in submodule necessary?

You mean in the test submodule?

**Comment by pickfire** [comment]
> You mean in the test submodule?

No, I mean library/stdarch and src/tools/cargo

**Comment by MarinPostma** [comment]
I'm note sure where these come from, I will remove them

**Comment by PhilipDaniels** [comment]
I am not an official reviewer so take this with a pinch of salt (from one who is also annoyed by the wordiness of the very frequent `Arc<Mutex<>>` pattern), but I would have thought if we were going to do this it would be better to follow the established `with_` pattern for specialist constructors, namely, add functions `Arc::with_mutex` and `RwLock::with_mutex`.

The way you have it at the moment feels a bit 'inside out'.

**Comment by MarinPostma** [comment]
@PhilipDaniels I had the `Box::pin` method in mind that returns a `Pin<Box>`. I think that when you want to use a mutex, you go look for `Mutex` and not for `Arc`.

File: library/std/src/sync/rwlock.rs (hunk 2 above)
Comment (pickfire) [inline library/std/src/sync/rwlock.rs:143]:
```suggestion
```

File: Cargo.lock [hunk 6]
```diff
@@ -126,9 +126,9 @@ dependencies = [
 
 [[package]]
 name = "autocfg"
-version = "1.0.0"
+version = "1.0.1"
```
Comment (jyn514) [inline Cargo.lock:129]: I don't think you meant to update this file.

File: library/std/src/lib.rs [hunk 7]
```diff
@@ -215,6 +215,7 @@
 // std is implemented with unstable features, many of which are internal
 // compiler details that will never be stable
 // NB: the following list is sorted to minimize merge conflicts.
+#![cfg_attr(not(bootstrap), feature(doc_spotlight))]
```
Comment (jyn514) [inline library/std/src/lib.rs:218]: This seems unrelated, maybe a rebase conflict?

Comment (MarinPostma) [reply library/std/src/lib.rs:218]: Yup it is, I was waiting for CI, will fix it as when I get to my computer

**Comment by JohnCSimon** [comment]
Ping from triage: 
@MarinPostma can you please address the merge conflicts? Thank you.

**Comment by JohnCSimon** [comment]
@MarinPostma unfortunately tests are failing

**Comment by JohnCSimon** [comment]
Pinging again from triage:
@MarinPostma unfortunately tests are failing

**Comment by Dylan-DPC** [comment]
@MarinPostma thanks for taking the time to contribute. I have to close this due to inactivity. If you wish and you have the time you can open a new PR with these changes and we'll take it from there. Thanks
